### ISSUE TITLE
Hangs in JPEGSegmentImageInputStream with bad data

### ISSUE THREAD

**Issue by marcreichman** [haraldk/TwelveMonkeys#510]
My project uses TwelveMonkeys (attempted the last 2 releases) to convert images into normalized formats (e.g. anything -> rgb jpeg) for later processing. We've encountered scenarios where junky data has hung while inside the `ImageIO.read` or the `ImageReader.getImageTypes` methods and never return. The major annoyance with this situation is that the input data is in a secure customer site and I do not even have access to it personally. Please note, these files are NOT usable, they don't work in browsers, windows paint, etc. I am not looking for a functioning load with them, but just for them not to hang. Without the TM library in classpath they simply do not load correctly with the default ImageIO paths but they don't hang.

I have put together an image testing harness using some similar code, and was able to get a trace of the threads at hang time. This is what is enabling me to write this ticket; unfortunately providing the actual input files is NOT going to be an option. I'm happy to provide this tester harness if you'd like, but again I can't provide the files.

This is the trace as I can provide it:
```
JPEGSegmentImageInputStream 279
JPEGSegmentImageInputStream 179
JPEGSegmentImageInputStream 387
JPEGSegmentImageInputStream 423
JPEGImageReader 312
JPEGImageReader 609
JPEGImageReader 347
JPEGImageReader 481
JPEGImageReader 602
JPEGImageReader 783
JPEGImageReader 773
JPEGImageReader 227
<my code>
ImageUtils 133
ImageUtils 61
```

Without the TM libraries loaded, the failure from the default ImageIO is an `IOException` with `Invalid JPEG file structure: two SOF markers`. This doesn't hang, just returns errors.

Happy to help with more info if necessary. Typically this has been tested with JDK 8u171, Linux x64 and Windows x64.

**Comment by haraldk** [comment]
Hi Marc,

Thanks for reporting!

I'd very much like to help, but it will be hard unless you can reproduce the problem in your test harness (with real or synthesized sample data). Can you edit it into one or more failing JUnit tests (preferably in `JPEGSegmentImageInputStreamTest`), that would help a lot!

For test data, note that we don't care so much about image data, what we need is the corrupted JIF (JPEG Interchange Format) "control" structure (SOI, SOF, APPn markers etc). So it should be possible to extract or create such data without exposing your clients private data.

I'll look into the stack trace to see if there are any obvious issues I can spot, but the more you can provide, the easier it will be, and the more likely a fix will happen soon. :-)


Best regards,

-- 
Harald K

**Comment by marcreichman** [comment]
Thanks Harald,

Any tips on how to either synthesize such data, or how to extract this from
the image data? It’s possible that nothing could leave the client site so
even with that it may not be possible to pull that data out.

**Comment by haraldk** [comment]
You could

* Start with one of the bad files, open it in a hex editor and delete everything from the scan (SOS) segment. If included, you probably also want to make sure you delete any XMP, Photoshop, Exif or IPTC metadata (I'm not sure what the images contain or what could potentially be compromised from this data, but just to make sure). 

* Start with random (good) JPEG file, open it in a hex editor and edit the structure to match the data that causes the hang. 

From the exception message from the built-in reader and the stack trace, the problem seems to related to the SOF (Start of Frame) and either multiple SOF markers in the file, or duplicate SOF ids in the SOF segment. 

Perhaps you could just send me contents of the SOF segment from some of the bad files?

See [JFIF on Wikipedia](https://en.wikipedia.org/wiki/JPEG_File_Interchange_Format) for more details on the format/structure. 


Hope that helps,

-- 
Harald K

**Comment by marcreichman** [comment]
Thanks for the feedback and ideas. I'll have to do some digging and see what we can do and get back to you.

Really appreciate the help and the library!

**Comment by haraldk** [comment]
Hi again Marc,

I looked at the stack trace and I may have found the issue. 

Can you try either checking out the source for version you use, or the latest master, and change line 279 in `JPEGSegmentInputStream` to:

                    while (componentIds.size() < 4 && !componentIds.add(id) && id < 255) {

This may potentially fix the problem. If it does, please report back, and I'll see if I can create a test.

-- 
Harald K

**Comment by marcreichman** [comment]
Hi Harald,

I will give it a shot and let you know. It may take a week or two for the turnaround. We've also been trying our luck with the hex editor to identify segment marker patterns in the "bad" files vs good ones.

In the meantime, given that you have a potential proposed fix, could you shed light on a pattern we could look for in the files to see if that's causing the issue?

**Comment by haraldk** [comment]
Yes,

You can look for duplicate or strange values for SOF or SOS component ids, duplicate SOF segments, more than 4 SOF or SOS components.

-- 
Harald K

**Comment by marcreichman** [comment]
Harald,

Thanks again for your continued attention and suggestions. Next time we're able to look at this in the secure network with the troubled files we'll run the patched code, and also try to find the patterns you mentioned, with a copy of jpegdump.c which should help print all this out, instead of using the hex editor. I expect this may take up to a week or two given the holiday season.

Thanks again and I will let you know!

Marc

**Comment by marcreichman** [comment]
Hi Harald!

Thanks for your patience. Your suggested fix does indeed eliminate the hang, so nice work on finding that given no input data! We are still working on identifying if there are obvious patterns in the header data, using a copy of `jpegdump`. I'll let you know if we find anything useful.

Thanks again for this library and for your help with this issue!

**Comment by haraldk** [comment]
Fixed in 5501c0e7094d094623648dc9355d9b620da7bef8
